File: VB/WebSite/App_Code/XtraReport1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI

''' <summary>
''' Summary description for XtraReport1
''' </summary>
Public Class XtraReport1
	Inherits DevExpress.XtraReports.UI.XtraReport
	Private Detail As DevExpress.XtraReports.UI.DetailBand
	Private dataSet11 As DataSet1
	Private xrPictureBox1 As XRPictureBox
	Private xrLabel2 As XRLabel
	Private xrLabel1 As XRLabel
	Private WithEvents DetailReport As DetailReportBand
	Private Detail1 As DetailBand
	Private xrLabel3 As XRLabel
	Private productsTableAdapter1 As DataSet1TableAdapters.ProductsTableAdapter
	Private categoriesTableAdapter1 As DataSet1TableAdapters.CategoriesTableAdapter
	Private WithEvents xrLabel4 As XRLabel
	Private xrLabel5 As XRLabel
	Private ReportHeader As ReportHeaderBand
	''' <summary>
	''' Required designer variable.
	''' </summary>
	Private components As System.ComponentModel.IContainer = Nothing

	Public Sub New()
		InitializeComponent()
		'
		' TODO: Add constructor logic here
		'
	End Sub

	''' <summary> 
	''' Clean up any resources being used.
	''' </summary>
	''' <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
	Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
		If disposing AndAlso (components IsNot Nothing) Then
			components.Dispose()
		End If
		MyBase.Dispose(disposing)
	End Sub

	#Region "Designer generated code"

	''' <summary>
	''' Required method for Designer support - do not modify
	''' the contents of this method with the code editor.
	''' </summary>
	Private Sub InitializeComponent()
		Dim resourceFileName As String = "XtraReport1.resx"
		Me.Detail = New DevExpress.XtraReports.UI.DetailBand()
		Me.xrLabel4 = New DevExpress.XtraReports.UI.XRLabel()
		Me.xrPictureBox1 = New DevExpress.XtraReports.UI.XRPictureBox()
		Me.xrLabel2 = New DevExpress.XtraReports.UI.XRLabel()
		Me.xrLabel1 = New DevExpress.XtraReports.UI.XRLabel()
		Me.xrLabel5 = New DevExpress.XtraReports.UI.XRLabel()
		Me.dataSet11 = New DataSet1()
		Me.DetailReport = New DevExpress.XtraReports.UI.DetailReportBand()
		Me.Detail1 = New DevExpress.XtraReports.UI.DetailBand()
		Me.xrLabel3 = New DevExpress.XtraReports.UI.XRLabel()
		Me.productsTableAdapter1 = New DataSet1TableAdapters.ProductsTableAdapter()
		Me.categoriesTableAdapter1 = New DataSet1TableAdapters.CategoriesTableAdapter()
		Me.ReportHeader = New DevExpress.XtraReports.UI.ReportHeaderBand()
		CType(Me.dataSet11, System.ComponentModel.ISupportInitialize).BeginInit()
		CType(Me, System.ComponentModel.ISupportInitialize).BeginInit()
		' 
		' Detail
		' 
		Me.Detail.Controls.AddRange(New DevExpress.XtraReports.UI.XRControl() { Me.xrLabel4, Me.xrPictureBox1, Me.xrLabel2, Me.xrLabel1})
		Me.Detail.Height = 69
		Me.Detail.Name = "Detail"
		Me.Detail.Padding = New DevExpress.XtraPrinting.PaddingInfo(0, 0, 0, 0, 100F)
		Me.Detail.TextAlignment = DevExpress.XtraPrinting.TextAlignment.TopLeft
		' 
		' xrLabel4
		' 
		Me.xrLabel4.BackColor = System.Drawing.Color.LightBlue
		Me.xrLabel4.Location = New System.Drawing.Point(0, 0)
		Me.xrLabel4.Name = "xrLabel4"
		Me.xrLabel4.Padding = New DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F)
		Me.xrLabel4.Size = New System.Drawing.Size(17, 17)
		Me.xrLabel4.StylePriority.UseBackColor = False
		Me.xrLabel4.Text = "-"
		Me.xrLabel4.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter
'		Me.xrLabel4.BeforePrint += New System.Drawing.Printing.PrintEventHandler(Me.xrLabel4_BeforePrint);
		' 
		' xrPictureBox1
		' 
		Me.xrPictureBox1.DataBindings.AddRange(New DevExpress.XtraReports.UI.XRBinding() { New DevExpress.XtraReports.UI.XRBinding("Image", Nothing, "Categories.Picture", "")})
		Me.xrPictureBox1.Location = New System.Drawing.Point(483, 0)
		Me.xrPictureBox1.Name = "xrPictureBox1"
		Me.xrPictureBox1.Size = New System.Drawing.Size(158, 58)
		Me.xrPictureBox1.Sizing = DevExpress.XtraPrinting.ImageSizeMode.ZoomImage
		' 
		' xrLabel2
		' 
		Me.xrLabel2.DataBindings.AddRange(New DevExpress.XtraReports.UI.XRBinding() { New DevExpress.XtraReports.UI.XRBinding("Text", Nothing, "Categories.Description", "")})
		Me.xrLabel2.Location = New System.Drawing.Point(75, 33)
		Me.xrLabel2.Name = "xrLabel2"
		Me.xrLabel2.Padding = New DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F)
		Me.xrLabel2.Size = New System.Drawing.Size(250, 25)
		Me.xrLabel2.Text = "xrLabel2"
		' 
		' xrLabel1
		' 
		Me.xrLabel1.DataBindings.AddRange(New DevExpress.XtraReports.UI.XRBinding() { New DevExpress.XtraReports.UI.XRBinding("Text", Nothing, "Categories.CategoryName", "")})
		Me.xrLabel1.Location = New System.Drawing.Point(75, 0)
		Me.xrLabel1.Name = "xrLabel1"
		Me.xrLabel1.Padding = New DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F)
		Me.xrLabel1.Size = New System.Drawing.Size(158, 25)
		Me.xrLabel1.Text = "xrLabel1"
		Me.xrLabel1.Font = New Font(Me.Font.FontFamily.ToString(), 12, FontStyle.Bold)
		' 
		' xrLabel5
		' 
		Me.xrLabel5.Font = New System.Drawing.Font("Times New Roman", 20.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, (CByte(204)))
		Me.xrLabel5.ForeColor = System.Drawing.Color.Blue
		Me.xrLabel5.Location = New System.Drawing.Point(275, 0)
		Me.xrLabel5.Name = "xrLabel5"
		Me.xrLabel5.Padding = New DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F)
		Me.xrLabel5.Size = New System.Drawing.Size(100, 33)
		Me.xrLabel5.StylePriority.UseFont = False
		Me.xrLabel5.StylePriority.UseForeColor = False
		Me.xrLabel5.Text = "Catalog"
		' 
		' dataSet11
		' 
		Me.dataSet11.DataSetName = "DataSet1"
		Me.dataSet11.SchemaSerializationMode = System.Data.SchemaSerializationMode.IncludeSchema
		' 
		' DetailReport
		' 
		Me.DetailReport.Bands.AddRange(New DevExpress.XtraReports.UI.Band() { Me.Detail1})
		Me.DetailReport.DataMember = "Products"
		Me.DetailReport.DataSource = Me.dataSet11
		Me.DetailReport.Level = 0
		Me.DetailReport.Name = "DetailReport"
'		Me.DetailReport.BeforePrint += New System.Drawing.Printing.PrintEventHandler(Me.DetailReport_BeforePrint);
		' 
		' Detail1
		' 
		Me.Detail1.Controls.AddRange(New DevExpress.XtraReports.UI.XRControl() { Me.xrLabel3})
		Me.Detail1.Height = 25
		Me.Detail1.Name = "Detail1"
		' 
		' xrLabel3
		' 
		Me.xrLabel3.DataBindings.AddRange(New DevExpress.XtraReports.UI.XRBinding() { New DevExpress.XtraReports.UI.XRBinding("Text", Nothing, "Products.ProductName", "")})
		Me.xrLabel3.Location = New System.Drawing.Point(225, 0)
		Me.xrLabel3.Name = "xrLabel3"
		Me.xrLabel3.Padding = New DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F)
		Me.xrLabel3.Size = New System.Drawing.Size(192, 25)
		Me.xrLabel3.Text = "xrLabel3"
		' 
		' productsTableAdapter1
		' 
		Me.productsTableAdapter1.ClearBeforeFill = True
		' 
		' categoriesTableAdapter1
		' 
		Me.categoriesTableAdapter1.ClearBeforeFill = True
		' 
		' ReportHeader
		' 
		Me.ReportHeader.Controls.AddRange(New DevExpress.XtraReports.UI.XRControl() { Me.xrLabel5})
		Me.ReportHeader.Height = 33
		Me.ReportHeader.Name = "ReportHeader"
		' 
		' XtraReport1
		' 
		Me.Bands.AddRange(New DevExpress.XtraReports.UI.Band() { Me.Detail, Me.DetailReport, Me.ReportHeader})
		Me.DataAdapter = Me.categoriesTableAdapter1
		Me.DataSource = Me.dataSet11
		Me.Version = "8.1"
		CType(Me.dataSet11, System.ComponentModel.ISupportInitialize).EndInit()
		CType(Me, System.ComponentModel.ISupportInitialize).EndInit()

	End Sub

	#End Region

	Private skippedDetailBands_Renamed As Hashtable

	Public Property SkippedDetailBands() As Hashtable
		Get
			If skippedDetailBands_Renamed Is Nothing Then
				skippedDetailBands_Renamed = New Hashtable()
			End If

			Return skippedDetailBands_Renamed
		End Get
		Set(ByVal value As Hashtable)
			skippedDetailBands_Renamed = value
		End Set
	End Property

	Public Sub UpdateDetail(ByVal detailID As Integer)
		If SkippedDetailBands.Contains(detailID) Then
			SkippedDetailBands(detailID) = Not Convert.ToBoolean(SkippedDetailBands(detailID))
		Else
			SkippedDetailBands.Add(detailID, True)
		End If

		Me.CreateDocument()
	End Sub

	Private Sub DetailReport_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles DetailReport.BeforePrint
		Dim catid As Integer = CInt(Fix(GetCurrentColumnValue("CategoryID")))

		' Fill the dataset with data from the data adapter.
		Me.productsTableAdapter1.Fill(Me.dataSet11.Products, catid)

		If SkippedDetailBands.Contains(catid) Then
			e.Cancel = Convert.ToBoolean(SkippedDetailBands(catid))
		End If
	End Sub

	Private Sub xrLabel4_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles xrLabel4.BeforePrint
		Dim label As XRLabel = TryCast(sender, XRLabel)
		Dim catid As Integer = CInt(Fix(GetCurrentColumnValue("CategoryID")))

		label.NavigateUrl = String.Format("javascript:DXReportViewerWindow.ASPxCallbackPanel1.PerformCallback({0})", catid.ToString())

		If SkippedDetailBands.Contains(catid) AndAlso CBool(SkippedDetailBands(catid)) = True Then
			label.Text = "+"
		Else
			label.Text = "-"
		End If
	End Sub

End Class

File: CS/WebSite/App_Code/XtraReport1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;

/// <summary>
/// Summary description for XtraReport1
/// </summary>
public class XtraReport1 : DevExpress.XtraReports.UI.XtraReport
{
    private DevExpress.XtraReports.UI.DetailBand Detail;
    private DataSet1 dataSet11;
    private XRPictureBox xrPictureBox1;
    private XRLabel xrLabel2;
    private XRLabel xrLabel1;
    private DetailReportBand DetailReport;
    private DetailBand Detail1;
    private XRLabel xrLabel3;
    private DataSet1TableAdapters.ProductsTableAdapter productsTableAdapter1;
    private DataSet1TableAdapters.CategoriesTableAdapter categoriesTableAdapter1;
    private XRLabel xrLabel4;
    private XRLabel xrLabel5;
    private ReportHeaderBand ReportHeader;
	/// <summary>
	/// Required designer variable.
	/// </summary>
	private System.ComponentModel.IContainer components = null;

	public XtraReport1()
	{
		InitializeComponent();
		//
		// TODO: Add constructor logic here
		//
	}
	
	/// <summary> 
	/// Clean up any resources being used.
	/// </summary>
	/// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
	protected override void Dispose(bool disposing) {
		if (disposing && (components != null)) {
			components.Dispose();
		}
		base.Dispose(disposing);
	}

	#region Designer generated code

	/// <summary>
	/// Required method for Designer support - do not modify
	/// the contents of this method with the code editor.
	/// </summary>
	private void InitializeComponent() {
        string resourceFileName = "XtraReport1.resx";
        this.Detail = new DevExpress.XtraReports.UI.DetailBand();
        this.xrLabel4 = new DevExpress.XtraReports.UI.XRLabel();
        this.xrPictureBox1 = new DevExpress.XtraReports.UI.XRPictureBox();
        this.xrLabel2 = new DevExpress.XtraReports.UI.XRLabel();
        this.xrLabel1 = new DevExpress.XtraReports.UI.XRLabel();
        this.xrLabel5 = new DevExpress.XtraReports.UI.XRLabel();
        this.dataSet11 = new DataSet1();
        this.DetailReport = new DevExpress.XtraReports.UI.DetailReportBand();
        this.Detail1 = new DevExpress.XtraReports.UI.DetailBand();
        this.xrLabel3 = new DevExpress.XtraReports.UI.XRLabel();
        this.productsTableAdapter1 = new DataSet1TableAdapters.ProductsTableAdapter();
        this.categoriesTableAdapter1 = new DataSet1TableAdapters.CategoriesTableAdapter();
        this.ReportHeader = new DevExpress.XtraReports.UI.ReportHeaderBand();
        ((System.ComponentModel.ISupportInitialize)(this.dataSet11)).BeginInit();
        ((System.ComponentModel.ISupportInitialize)(this)).BeginInit();
        // 
        // Detail
        // 
        this.Detail.Controls.AddRange(new DevExpress.XtraReports.UI.XRControl[] {
            this.xrLabel4,
            this.xrPictureBox1,
            this.xrLabel2,
            this.xrLabel1});
        this.Detail.Height = 69;
        this.Detail.Name = "Detail";
        this.Detail.Padding = new DevExpress.XtraPrinting.PaddingInfo(0, 0, 0, 0, 100F);
        this.Detail.TextAlignment = DevExpress.XtraPrinting.TextAlignment.TopLeft;
        // 
        // xrLabel4
        // 
        this.xrLabel4.BackColor = System.Drawing.Color.Red;
        this.xrLabel4.Location = new System.Drawing.Point(0, 0);
        this.xrLabel4.Name = "xrLabel4";
        this.xrLabel4.Padding = new DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F);
        this.xrLabel4.Size = new System.Drawing.Size(17, 17);
        this.xrLabel4.StylePriority.UseBackColor = false;
        this.xrLabel4.Text = "-";
        this.xrLabel4.BeforePrint += new System.Drawing.Printing.PrintEventHandler(this.xrLabel4_BeforePrint);
        // 
        // xrPictureBox1
        // 
        this.xrPictureBox1.DataBindings.AddRange(new DevExpress.XtraReports.UI.XRBinding[] {
            new DevExpress.XtraReports.UI.XRBinding("Image", null, "Categories.Picture", "")});
        this.xrPictureBox1.Location = new System.Drawing.Point(483, 0);
        this.xrPictureBox1.Name = "xrPictureBox1";
        this.xrPictureBox1.Size = new System.Drawing.Size(158, 58);
        this.xrPictureBox1.Sizing = DevExpress.XtraPrinting.ImageSizeMode.ZoomImage;
        // 
        // xrLabel2
        // 
        this.xrLabel2.DataBindings.AddRange(new DevExpress.XtraReports.UI.XRBinding[] {
            new DevExpress.XtraReports.UI.XRBinding("Text", null, "Categories.Description", "")});
        this.xrLabel2.Location = new System.Drawing.Point(75, 33);
        this.xrLabel2.Name = "xrLabel2";
        this.xrLabel2.Padding = new DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F);
        this.xrLabel2.Size = new System.Drawing.Size(250, 25);
        this.xrLabel2.Text = "xrLabel2";
        // 
        // xrLabel1
        // 
        this.xrLabel1.DataBindings.AddRange(new DevExpress.XtraReports.UI.XRBinding[] {
            new DevExpress.XtraReports.UI.XRBinding("Text", null, "Categories.CategoryName", "")});
        this.xrLabel1.Location = new System.Drawing.Point(75, 0);
        this.xrLabel1.Name = "xrLabel1";
        this.xrLabel1.Padding = new DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F);
        this.xrLabel1.Size = new System.Drawing.Size(158, 25);
        this.xrLabel1.Text = "xrLabel1";
        // 
        // xrLabel5
        // 
        this.xrLabel5.Font = new System.Drawing.Font("Times New Roman", 20.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ((byte)(204)));
        this.xrLabel5.ForeColor = System.Drawing.Color.Blue;
        this.xrLabel5.Location = new System.Drawing.Point(275, 0);
        this.xrLabel5.Name = "xrLabel5";
        this.xrLabel5.Padding = new DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F);
        this.xrLabel5.Size = new System.Drawing.Size(100, 33);
        this.xrLabel5.StylePriority.UseFont = false;
        this.xrLabel5.StylePriority.UseForeColor = false;
        this.xrLabel5.Text = "Catalog";
        // 
        // dataSet11
        // 
        this.dataSet11.DataSetName = "DataSet1";
        this.dataSet11.SchemaSerializationMode = System.Data.SchemaSerializationMode.IncludeSchema;
        // 
        // DetailReport
        // 
        this.DetailReport.Bands.AddRange(new DevExpress.XtraReports.UI.Band[] {
            this.Detail1});
        this.DetailReport.DataMember = "Products";
        this.DetailReport.DataSource = this.dataSet11;
        this.DetailReport.Level = 0;
        this.DetailReport.Name = "DetailReport";
        this.DetailReport.BeforePrint += new System.Drawing.Printing.PrintEventHandler(this.DetailReport_BeforePrint);
        // 
        // Detail1
        // 
        this.Detail1.Controls.AddRange(new DevExpress.XtraReports.UI.XRControl[] {
            this.xrLabel3});
        this.Detail1.Height = 25;
        this.Detail1.Name = "Detail1";
        // 
        // xrLabel3
        // 
        this.xrLabel3.DataBindings.AddRange(new DevExpress.XtraReports.UI.XRBinding[] {
            new DevExpress.XtraReports.UI.XRBinding("Text", null, "Products.ProductName", "")});
        this.xrLabel3.Location = new System.Drawing.Point(225, 0);
        this.xrLabel3.Name = "xrLabel3";
        this.xrLabel3.Padding = new DevExpress.XtraPrinting.PaddingInfo(2, 2, 0, 0, 100F);
        this.xrLabel3.Size = new System.Drawing.Size(192, 25);
        this.xrLabel3.Text = "xrLabel3";
        // 
        // productsTableAdapter1
        // 
        this.productsTableAdapter1.ClearBeforeFill = true;
        // 
        // categoriesTableAdapter1
        // 
        this.categoriesTableAdapter1.ClearBeforeFill = true;
        // 
        // ReportHeader
        // 
        this.ReportHeader.Controls.AddRange(new DevExpress.XtraReports.UI.XRControl[] {
            this.xrLabel5});
        this.ReportHeader.Height = 33;
        this.ReportHeader.Name = "ReportHeader";
        // 
        // XtraReport1
        // 
        this.Bands.AddRange(new DevExpress.XtraReports.UI.Band[] {
            this.Detail,
            this.DetailReport,
            this.ReportHeader});
        this.DataAdapter = this.categoriesTableAdapter1;
        this.DataSource = this.dataSet11;
        this.Version = "8.1";
        ((System.ComponentModel.ISupportInitialize)(this.dataSet11)).EndInit();
        ((System.ComponentModel.ISupportInitialize)(this)).EndInit();

	}

	#endregion

    private Hashtable skippedDetailBands;

    public Hashtable SkippedDetailBands
    {
        get {
            if (skippedDetailBands == null)
                skippedDetailBands = new Hashtable();

            return skippedDetailBands;
        }
        set { skippedDetailBands = value; }
    }

    public void UpdateDetail(int detailID)
    {
        if (SkippedDetailBands.Contains(detailID))
            SkippedDetailBands[detailID] = !Convert.ToBoolean(SkippedDetailBands[detailID]);
        else
            SkippedDetailBands.Add(detailID, true);

        this.CreateDocument();
    }

    private void DetailReport_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
    {
        int catid = (int)GetCurrentColumnValue("CategoryID");

        // Fill the dataset with data from the data adapter.
        this.productsTableAdapter1.Fill(this.dataSet11.Products, catid);
 
        if (SkippedDetailBands.Contains(catid))
            e.Cancel = Convert.ToBoolean(SkippedDetailBands[catid]);
    }

    private void xrLabel4_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
    {    
        XRLabel label = sender as XRLabel;
        int catid = (int)GetCurrentColumnValue("CategoryID");

        label.NavigateUrl = string.Format("javascript:ASPxCallbackPanel1.PerformCallback({0})", catid.ToString());

        if (SkippedDetailBands.Contains(catid) && (bool)SkippedDetailBands[catid] == true)
            label.Text = "+";
        else
            label.Text = "-";
    }

}
